src/prefect/server/api/concurrency_limits.py

```diff
@@ -188,6 +474,103 @@ async def increment_concurrency_limits_v1(
     ),
     db: PrefectDBInterface = Depends(provide_database_interface),
 ) -> List[MinimalConcurrencyLimitResponse]:
+    # Only use V2 path when all tag: limits exist; else fall back to V1
+    v2_names = [f"tag:{tag}" for tag in names]
+    async with db.session_context(begin_transaction=True) as session:
+        existing_limits = await cl_v2_models.bulk_read_concurrency_limits(
+            session=session, names=v2_names
+        )
+
+        if existing_limits and len(existing_limits) == len(v2_names):
+            zero = next(
+                (limit for limit in existing_limits if getattr(limit, "limit", 0) == 0),
+                None,
+            )
+            if zero is not None:
+                zero_tag = str(zero.name).removeprefix("tag:")
+                raise HTTPException(
+                    status_code=status.HTTP_423_LOCKED,
+                    detail=(
+                        f'The concurrency limit on tag "{zero_tag}" is 0 and will '
+                        "deadlock if the task tries to run again."
+                    ),
+                )
+
+            active_limit_ids = [
+                limit.id for limit in existing_limits if bool(limit.active)
+            ]
+            acquired = False
+            if active_limit_ids:
+                acquired = await cl_v2_models.bulk_increment_active_slots(
+                    session=session,
+                    concurrency_limit_ids=active_limit_ids,
+                    slots=1,
+                )
+
+    if existing_limits and len(existing_limits) == len(v2_names):
```

Comment:
We want to get v1 and v2 limits and increment each of them. Limits will stop working if you have mixed v1 and v2 limits given.